ledger/object/index_db.go

```diff
@@ -17,75 +17,98 @@
 	"go.opencensus.io/stats"
 )
 
 // IndexDB is a db-based storage, that stores a collection of IndexBuckets
 type IndexDB struct {
-	lock sync.RWMutex
-	db   store.DB
-
+	pool        *pgxpool.Pool
 	recordStore *RecordDB
 }
 
-type indexKey struct {
-	pn    insolar.PulseNumber
-	objID insolar.ID
+var ReadTxOptions = pgx.TxOptions{
+	IsoLevel:       pgx.Serializable,
+	AccessMode:     pgx.ReadOnly,
+	DeferrableMode: pgx.NotDeferrable,
 }
 
-func newIndexKey(raw []byte) indexKey {
-	ik := indexKey{}
-	ik.pn = insolar.NewPulseNumber(raw)
-	ik.objID = *insolar.NewIDFromBytes(raw[ik.pn.Size():])
-
-	return ik
+var WriteTxOptions = pgx.TxOptions{
+	IsoLevel:       pgx.Serializable,
+	AccessMode:     pgx.ReadWrite,
+	DeferrableMode: pgx.NotDeferrable,
 }
 
-func (k indexKey) Scope() store.Scope {
-	return store.ScopeIndex
+// NewIndexDB creates a new instance of IndexDB
+func NewIndexDB(pool *pgxpool.Pool, records *RecordDB) *IndexDB {
+	return &IndexDB{pool: pool, recordStore: records}
 }
 
-func (k indexKey) ID() []byte {
-	return append(k.pn.Bytes(), k.objID.Bytes()...)
-}
+// SetIndex adds a bucket with provided pulseNumber and ID
+func (i *IndexDB) SetIndex(ctx context.Context, pn insolar.PulseNumber, bucket record.Index) error {
+	conn, err := i.pool.Acquire(ctx)
+	if err != nil {
+		return errors.Wrap(err, "Unable to acquire a database connection")
+	}
+	defer conn.Release()
 
-type lastKnownIndexPNKey struct {
-	objID insolar.ID
-}
+	log := inslogger.FromContext(ctx)
 
-func (k lastKnownIndexPNKey) Scope() store.Scope {
-	return store.ScopeLastKnownIndexPN
-}
+	for { // retry loop
+		tx, err := conn.BeginTx(ctx, WriteTxOptions)
+		if err != nil {
+			return errors.Wrap(err, "Unable to start a write transaction")
+		}
 
-func (k lastKnownIndexPNKey) ID() []byte {
-	id := k.objID
-	return bytes.Join([][]byte{id.Pulse().Bytes(), id.Hash()}, nil)
-}
+		var pendingRecords [][]byte
+		for _, pr := range bucket.PendingRecords {
+			pendingRecords = append(pendingRecords, pr.Bytes())
+		}
+		var latestState []byte
+		if bucket.Lifeline.LatestState != nil {
+			latestState = bucket.Lifeline.LatestState.Bytes()
+		}
+		var latestRequest []byte
+		if bucket.Lifeline.LatestRequest != nil {
+			latestRequest = bucket.Lifeline.LatestRequest.Bytes()
+		}
 
-// NewIndexDB creates a new instance of IndexDB
-func NewIndexDB(db store.DB, recordStore *RecordDB) *IndexDB {
-	return &IndexDB{db: db, recordStore: recordStore}
-}
+		_, err = tx.Exec(ctx, `
+			INSERT INTO indexes(object_id, pulse_number, lifeline_last_used, pending_records,
+				latest_state, state_id, parent, latest_request, earliest_open_request, open_requests_count)
+			VALUES($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
+			ON CONFLICT (object_id, pulse_number)
+			DO UPDATE SET 
+				lifeline_last_used = $3,
+				pending_records = $4,
+				latest_state = $5,
+				state_id = $6,
+				parent = $7,
+				latest_request = $8,
+				earliest_open_request = $9,
+				open_requests_count = $10
```

Comment:
you can use `EXCLUDED.lifeline_last_used` instead of `$3`, etc. This will result in shorter query and lesser parse time.